### PR TITLE
Allow for transactions to be renamed using email

### PR DESCRIPTION
**Pull request by sampoder**
## Summary of the problem

Request from Zach here: https://hackclub.slack.com/archives/CN523HLKW/p1704482178687159

## Describe your changes

We parse the email's text to check if there's a line starting with "Rename to:" and use that line to rename the transaction.

## Before merging

I'd like feedback on:

1. Making the logic in the mailbox cleaner
2. Improving the copy on the initial email sent to users after the transaction is approved
3. Does anyone see any edge cases?

### CODE REVIEW

**Comment by sampoder** [comment]
Oh and if there's a better thing someone has in mind than Rename to: that'd be awesome. It's still a little jank.

I'll fix linting when I get home.

**Comment by northeastprince** [comment]
I like Zach's suggestion of taking what's in between quotation marks if it's the only thing in the body. We could also rename it based on the subject in the reply.

**Comment by sampoder** [comment]
@northeastprince: whipped up an alternative here: https://github.com/hackclub/hcb/pull/4931!

**Comment by garyhtou** [comment]
What are your thoughts on Helpscout's style?
https://docs.helpscout.com/article/67-respond-to-email-notifications-to-take-action-in-help-scout#ecommands

```
@memo 📦 10 boxes shipped via USPS
```

Regex: `/(^|\s)(?<tag>@memo)\s+(?<content>.+)/i`

File: app/views/canonical_pending_transaction_mailer/notify_approved.html.erb [hunk 1]
```diff
@@ -4,6 +4,8 @@ A charge of <%= render_money @cpt.amount_cents %> at '<%= @cpt.memo %>' occurred
   <%= link_to "Upload your receipt for this charge. (this link can used by anyone for two weeks)", @upload_url %>
 
   <% if @receipt_upload_feature %>
-    Alternatively, <em>reply to this email with an image or PDF receipt attached</em>.
+    <p>
+      Alternatively, <em>reply to this email with an image or PDF receipt attached</em>. Include a line beginning with <em>Rename to:</em> to rename the transaction.
```
Comment (sampoder) [inline app/views/canonical_pending_transaction_mailer/notify_approved.html.erb]:
```suggestion
      Alternatively, <em>reply to this email with an image or PDF receipt attached</em>. Include a line beginning with <em>@rename</em> to rename the transaction.
```

File: app/mailboxes/hcb_code_receipts_mailbox.rb [hunk 2]
```diff
@@ -28,9 +28,27 @@ def process
...
+           else
+             mail.body.decoded
+           end
+
+    custom_memo = nil
+
+    rename_to = text&.lines&.find { |line| line.start_with?("Rename to:") }
+
+    if rename_to
+      rename_to.slice!("Rename to:")
```
Comment (sampoder) [inline app/mailboxes/hcb_code_receipts_mailbox.rb]:
```suggestion
      rename_to.slice!("@rename")
```

File: app/mailboxes/hcb_code_receipts_mailbox.rb [hunk 3]
```diff
@@ -28,9 +28,27 @@ def process
...
 
+    text = if mail.multipart?
+             mail.text_part&.body&.decoded
+           else
+             mail.body.decoded
+           end
+
+    custom_memo = nil
+
+    rename_to = text&.lines&.find { |line| line.start_with?("Rename to:") }
```
Comment (sampoder) [inline app/mailboxes/hcb_code_receipts_mailbox.rb]:
```suggestion
    rename_to = text&.lines&.find { |line| line.start_with?("@rename") }
```

File: app/mailboxes/hcb_code_receipts_mailbox.rb [hunk 4]
```diff
@@ -28,9 +28,27 @@ def process
 
     return bounce_error if result.empty?
 
+    text = if mail.multipart?
+             mail.text_part&.body&.decoded
+           else
+             mail.body.decoded
+           end
```
Comment (garyhtou) [inline app/mailboxes/hcb_code_receipts_mailbox.rb]:
```suggestion
    content = text || body || html
```

File: app/views/hcb_code_receipts_mailer/bounce_success.html.erb [hunk 5]
```diff
@@ -1,5 +1,9 @@
 <p>
-We got your <%= "receipt".pluralize(@receipts_count) %>, thanks!
+  <% if @renamed_to %>
+    We got your <%= "receipt".pluralize(@receipts_count) %> and renamed the transaction to <i><%= @renamed_to %></i>, thanks!
```
Comment (garyhtou) [inline app/views/hcb_code_receipts_mailer/bounce_success.html.erb]:
```suggestion
    We got your <%= "receipt".pluralize(@receipts_count) %> and renamed the transaction to <i>"<%= @renamed_to %>"</i>, thanks!
```

File: app/views/canonical_pending_transaction_mailer/notify_approved.html.erb [hunk 6]
```diff
@@ -4,6 +4,8 @@ A charge of <%= render_money @cpt.amount_cents %> at '<%= @cpt.memo %>' occurred
   <%= link_to "Upload your receipt for this charge. (this link can used by anyone for two weeks)", @upload_url %>
 
   <% if @receipt_upload_feature %>
-    Alternatively, <em>reply to this email with an image or PDF receipt attached</em>.
+    <p>
+      Alternatively, <em>reply to this email with an image or PDF receipt attached</em>. Include a line beginning with <em>@rename</em> to rename the transaction.
```
Comment (sampoder) [inline app/views/canonical_pending_transaction_mailer/notify_approved.html.erb]:
```suggestion
      Alternatively, <em>reply to this email with an image or PDF receipt attached</em>. When responding with a receipt, include a line beginning with <em>@rename</em> to rename the transaction.
```

File: app/mailboxes/hcb_code_receipts_mailbox.rb [hunk 7]
```diff
@@ -28,9 +28,27 @@ def process
...
+
+    custom_memo = nil
+
+    rename_to = text&.lines&.find { |line| line.start_with?("@rename") }
+
+    if rename_to
+      rename_to.slice!("@rename")
+      custom_memo = rename_to.strip
```
Comment (garyhtou) [inline app/mailboxes/hcb_code_receipts_mailbox.rb]:
```suggestion
    email_command = content.match(/(^|\s)(?<tag>@rename)\s+(?<content>.+)/i)
    custom_memo = email_command&.[]("content")&.strip

    if custom_memo
```

**Review by garyhtou — APPROVED** [review]
yay!

File: app/mailboxes/hcb_code_receipts_mailbox.rb [hunk 8]
```diff
@@ -28,9 +28,20 @@ def process
...
     return bounce_error if result.empty?
 
+    content = text || body || html
+
+    email_command = content.match(/(^|\s)(?<tag>@rename)\s+(?<content>.+)/i)
+    custom_memo = email_command&.[]("content")&.strip
+
+    if custom_memo
+      @hcb_code.canonical_transactions.each { |ct| ct.update!(custom_memo:) }
+      @hcb_code.canonical_pending_transactions.each { |cpt| cpt.update!(custom_memo:) }
```
Comment (northeastprince) [inline app/mailboxes/hcb_code_receipts_mailbox.rb]: In what circumstances would there be multiple?

Comment (garyhtou) [reply app/mailboxes/hcb_code_receipts_mailbox.rb]: fee reimbursements!

and maybe also stripe issuing multi captures? (i don't remember if they have separate hcb codes)

**Pull request merged**